src/shared/telemetry/recorder.ts

```diff
@@ -0,0 +1,189 @@
+): U & TelemetryRecorder<T> {
+    function uncapitalize(s: string): string {
+        return `${s[0].toLowerCase()}${s.slice(1)}`
+    }
+
+    const proxy = new Proxy(inner, {
+        get: (target, prop, receiver) => {
+            if (typeof prop !== 'string') {
+                throw new TypeError(`Invalid telemetry recorder key: ${String(prop)}`)
+            }
+
+            const name = uncapitalize(prop.replace('record', ''))
+
+            if (!prop.startsWith('record')) {
+                // We should find the last event with the same name but this works for now
+                return Reflect.get(target, prop, receiver) ?? queue.find(e => e.name === name)
+            }
+
+            return <U>(value: U) => {
+                queue.push({ value, source, name, timestamp: new Date() })
+                Reflect.set(target, name, value, receiver)
+
+                return proxy
+            }
+        },
+    }) as U & TelemetryRecorder<T>
+
+    return proxy
+}
+
+function startMetric<T>(id: string, record: (data: T) => void): ActiveMetric<T> {
+    const queue = storage.getStore()?.telemetry.queue ?? []
+    const createTime = new Date()
+
+    function stop() {
+        const result: Record<string, any> = { createTime }
+        result.duration = Date.now() - createTime.getTime()
+
+        for (const event of queue.filter(event => event.source === id)) {
+            result[event.name] = event.value
+        }
+
+        // We are currently unable to do any validation here as it is implemented
+        // inside the generated `record...` functions.
+        record(result as T)
+    }
+
+    function has(this: Partial<T>, ...props: (keyof T)[]): boolean {
+        return props.map(p => this[p] !== undefined).reduce((a, b) => a && b, true)
+    }
+
+    return createRecorder({ stop, record, has }, id, queue) as ActiveMetric<T>
+}
+
+type Telemetry = Omit<typeof telemetry, 'millisecondsSince'>
+type RemoveRecord<T> = T extends `record${infer U}` ? U : T
+type Metadata<P extends keyof Telemetry> = NonNullable<Parameters<Telemetry[P]>[0]>
+
+type MappedMetrics = {
+    [P in keyof Telemetry as RemoveRecord<P>]: Metric<Metadata<P>>
+}
+
+const metrics = {} as MappedMetrics
+
+for (const [k, v] of Object.entries(telemetry)) {
+    const name = k.replace('record', '') as keyof MappedMetrics
+    const start = () => startMetric<any>(name, v)
+    metrics[name] = { start } as any
+}
+
+interface ExecutionContext {
+    readonly name: string
+    readonly logger: Logger
+    readonly telemetry: { readonly queue: TelemetryEvent[] }
+}
+
+const storage = new AsyncLocalStorage<ExecutionContext>()
+
+/**
+ * Runs a function with an {@link ExecutionContext}, containing context-specific observability
+ * utilities such as telemetry and logging.
+ */
+export function run<F extends (...args: any[]) => any>(name: string, fn: F, ...args: Parameters<F>): ReturnType<F> {
+    const context = {
+        name,
+        logger: getLogger(),
```

Comment:
We could also associate log metadata with specific executions, though I didn't implement anything.